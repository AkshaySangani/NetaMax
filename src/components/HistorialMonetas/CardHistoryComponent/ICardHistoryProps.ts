export interface ICardHistoryProps {
  /**
   * The detail card history card
   * @type {string}
   */
  detail: string;

  /**
   * The amount card history card
   * @type {number}
   */
  amount: number;

  /**
   * The date card history card
   * @type {Date}
   */
  date: Date;

  /**
   * The orderId card history card
   * @type {number}
   */
  orderId?: number;

  /**
   * The movementTypeId card history card
   * @type {number}
   */
  movementTypeId: number;
}
