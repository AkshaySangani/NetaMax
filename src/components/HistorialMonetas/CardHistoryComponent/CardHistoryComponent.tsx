import React, { ReactElement } from "react";

import { MIS_COMPRAS_ROUTE } from "constants/checkoutConstants";
import { CURRENCY_NAME } from "constants/localeFormatConstants";
import ArrowFullTop from "styled/icons/ArrowFullTop";
import Clock from "styled/icons/Clock";
import Order from "styled/icons/Order";
import { formatMoney } from "utils/currencyUtils";

import { Box, Text } from "@chakra-ui/react";

import { ICardHistoryProps } from "./ICardHistoryProps";

/**
 * Card History Component
 * @param {ICardHistoryProps} props the component props.
 * @returns {ReactElement} the component element.
 */
const CardHistoryComponent = (props: ICardHistoryProps): ReactElement => {
  const { detail, amount, date, orderId, movementTypeId } = props;
  const [cardTypeColor, setCardTypeColor] = React.useState("");
  const [cardTypeReason, setCardTypeReason] = React.useState("");
  const dateFormat = new Date(date);

  /**
   * get difference date to date card history
   * @returns {string} the date ago.
   */
  const getDifferenceDate = (): string => {
    const dateNow = new Date();
    const differenceDate = dateNow.getTime() - dateFormat.getTime();

    const differenceHours = Math.floor(differenceDate / (1000 * 3600));
    const differenceDays = Math.floor(differenceHours / 24);

    if (differenceHours < 1) {
      return "Hace menos de 1 hora";
    }
    if (differenceHours < 24) {
      return `Hace ${differenceHours} ${differenceHours > 1 ? "horas" : "hora"}`;
    }
    if (differenceDays < 7) {
      return `Hace ${differenceDays} ${differenceDays > 1 ? "días" : "día"}`;
    }

    const options = {
      year: "numeric",
      month: "long",
      day: "numeric",
    } as Intl.DateTimeFormatOptions;

    return dateFormat.toLocaleDateString("es-MX", options);
  };

  React.useEffect(() => {
    switch (movementTypeId) {
      case 10:
        setCardTypeColor("#04A940");
        setCardTypeReason("Abono");
        break;
      case 20:
        setCardTypeColor("#ED9F00");
        setCardTypeReason("Usados");
        break;
      case 30:
        setCardTypeColor("#F54A49");
        setCardTypeReason("Vencidos");
        break;
    }
  }, [movementTypeId]);

  return (
    <Box
      borderWidth="1px"
      borderStyle="solid"
      borderColor="#9F9F9F"
      borderRadius="14px"
      padding="1rem"
      display="flex"
      flexDirection="row"
      justifyContent="space-between"
      width="100%"
      position="relative"
    >
      <Box display="flex" flexDirection="column" flex="1">
        <Text fontSize="16px" color="#2F3346" fontWeight="600">
          {detail}
        </Text>
        <Box display="flex" flexDirection="row" placeItems="center">
          <Box display="flex" flexDirection="row" alignItems="center">
            <Box
              width="18px"
              height="18px"
              display="flex"
              flexDirection="column"
              alignItems="center"
              justifyContent="center"
            >
              <Clock display="block" />
            </Box>
            <Box display="flex" flexDirection="column">
              <Text
                data-testid="dateTest"
                fontSize="14px"
                color="#717171"
                paddingLeft="0.25rem"
                lineHeight="normal"
              >
                {getDifferenceDate()}
              </Text>
            </Box>
          </Box>
          {orderId && (
            <Box display="flex" flexDirection="row" paddingLeft="0.5rem" alignItems="center">
              <Box
                width="18px"
                height="18px"
                display="flex"
                flexDirection="column"
                alignItems="center"
                justifyContent="center"
              >
                <Order display="block" />
              </Box>
              <Box display="flex" flexDirection="column">
                <a href={`${MIS_COMPRAS_ROUTE}/${orderId}`}>
                  <Text
                    fontSize="14px"
                    color="#3A86FF"
                    paddingLeft="0.25rem"
                    lineHeight="normal"
                    textDecoration="underline"
                    cursor="pointer"
                  >
                    Detalle de compra
                  </Text>
                </a>
              </Box>
            </Box>
          )}
        </Box>
      </Box>
      <Box display="flex" flexDirection="column" alignItems="end">
        <Box display="flex" flexDirection="row" placeItems="center" lineHeight="24px">
          <Text
            data-testid="amountTest"
            fontSize="18px"
            color={cardTypeColor}
            fontWeight="700"
            paddingRight="0.25rem"
          >
            {movementTypeId === 10 ? "+" : "-"} {formatMoney(Number(amount), CURRENCY_NAME)}
          </Text>
          <ArrowFullTop
            fill={cardTypeColor}
            transform={movementTypeId != 10 ? "rotate(180)" : ""}
          />
        </Box>
        <Box display="flex" flexDirection="row" placeItems="center">
          <Text fontSize="14px" color="#717171" paddingLeft="0.25rem">
            {cardTypeReason}
          </Text>
        </Box>
      </Box>
      <Box
        position="absolute"
        width="6px"
        backgroundColor={cardTypeColor}
        height="3em"
        right="-1px"
      />
    </Box>
  );
};

export default CardHistoryComponent;
