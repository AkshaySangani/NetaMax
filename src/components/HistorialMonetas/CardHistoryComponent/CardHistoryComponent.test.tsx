import { render } from "@testing-library/react";
import CardHistoryComponent from "./CardHistoryComponent";
import { ICardHistoryProps } from "./ICardHistoryProps";

jest.mock("next/router", () => require("next-router-mock"));

describe("Render CardHistory component", () => {
  const defaultProps: ICardHistoryProps = {
    detail: "detail",
    amount: 1,
    date: new Date(),
    movementTypeId: 10,
    orderId: 123,
  };

  it("CardHistoryComponent should be render correctly", () => {
    const { container } = render(<CardHistoryComponent {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  describe("movement Type test", () => {
    const movementTypeIds = [
      [10, "+ $1.00"],
      [20, "- $1.00"],
      [30, "- $1.00"],
    ];

    it.each(movementTypeIds)(
      "CardHistoryComponent should be render correctly, when movementTypeId is %p",
      (typeId, value) => {
        const { getByTestId } = render(
          <CardHistoryComponent {...defaultProps} movementTypeId={parseInt(typeId.toString())} />
        );
        expect(getByTestId("amountTest").textContent).toBe(value);
      }
    );
  });

  describe("date test", () => {
    const options = {
      year: "numeric",
      month: "long",
      day: "numeric",
    } as Intl.DateTimeFormatOptions;

    const monthAgo = new Date(new Date().setMonth(new Date().getMonth() - 1));

    const dates = [
      [new Date(), "Hace menos de 1 hora"],
      [new Date(new Date().setHours(new Date().getHours() - 1)), "Hace 1 hora"],
      [new Date(new Date().setHours(new Date().getHours() - 3)), "Hace 3 horas"],
      [new Date(new Date().setHours(new Date().getHours() - 25)), "Hace 1 día"],
      [new Date(new Date().setHours(new Date().getHours() - 49)), "Hace 2 días"],
      [monthAgo, monthAgo.toLocaleDateString("es-MX", options)],
    ];

    it.each(dates)(
      "CardHistoryComponent should be render correctly, when movementTypeId is %p",
      (date, value) => {
        const { getByTestId } = render(
          <CardHistoryComponent {...defaultProps} date={date as Date} />
        );
        expect(getByTestId("dateTest").textContent).toBe(value);
      }
    );
  });
});
