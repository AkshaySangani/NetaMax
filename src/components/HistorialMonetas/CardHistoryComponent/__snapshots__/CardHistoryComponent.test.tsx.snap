// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render CardHistory component CardHistoryComponent should be render correctly 1`] = `
<div>
  <div
    class="css-1cac0wi"
  >
    <div
      class="css-b95f0i"
    >
      <p
        class="chakra-text css-1ob8n37"
      >
        detail
      </p>
      <div
        class="css-1xudhsy"
      >
        <div
          class="css-u4p24i"
        >
          <div
            class="css-pe2mn1"
          >
            <svg
              display="block"
              fill="none"
              height="17"
              viewBox="0 0 16 17"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 5.5V8.5L10.25 10.75M14.75 8.5C14.75 12.2279 11.7279 15.25 8 15.25C4.27208 15.25 1.25 12.2279 1.25 8.5C1.25 4.77208 4.27208 1.75 8 1.75C11.7279 1.75 14.75 4.77208 14.75 8.5Z"
                stroke="#2F3346"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </div>
          <div
            class="css-j7qwjs"
          >
            <p
              class="chakra-text css-1pn73qc"
              data-testid="dateTest"
            >
              Hace menos de 1 hora
            </p>
          </div>
        </div>
        <div
          class="css-1pt6qwo"
        >
          <div
            class="css-pe2mn1"
          >
            <svg
              display="block"
              fill="none"
              height="18"
              viewBox="0 0 15 18"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.83276 9.87623C5.44223 9.4857 4.80907 9.4857 4.41854 9.87623C4.02802 10.2668 4.02802 10.8999 4.41854 11.2904L5.83276 9.87623ZM6.70898 12.1667L6.00188 12.8738C6.3924 13.2643 7.02557 13.2643 7.41609 12.8738L6.70898 12.1667ZM10.5828 9.70711C10.9733 9.31658 10.9733 8.68342 10.5828 8.29289C10.1922 7.90237 9.55907 7.90237 9.16854 8.29289L10.5828 9.70711ZM12.0423 5.04167V14.5417H14.0423V5.04167H12.0423ZM11.459 15.125H3.54232V17.125H11.459V15.125ZM2.95898 14.5417V5.04167H0.958984V14.5417H2.95898ZM3.54232 4.45833H5.12565V2.45833H3.54232V4.45833ZM9.87565 4.45833H11.459V2.45833H9.87565V4.45833ZM3.54232 15.125C3.22015 15.125 2.95898 14.8638 2.95898 14.5417H0.958984C0.958984 15.9684 2.11558 17.125 3.54232 17.125V15.125ZM12.0423 14.5417C12.0423 14.8638 11.7812 15.125 11.459 15.125V17.125C12.8857 17.125 14.0423 15.9684 14.0423 14.5417H12.0423ZM14.0423 5.04167C14.0423 3.61493 12.8857 2.45833 11.459 2.45833V4.45833C11.7812 4.45833 12.0423 4.7195 12.0423 5.04167H14.0423ZM2.95898 5.04167C2.95898 4.7195 3.22015 4.45833 3.54232 4.45833V2.45833C2.11558 2.45833 0.958984 3.61493 0.958984 5.04167H2.95898ZM4.41854 11.2904L6.00188 12.8738L7.41609 11.4596L5.83276 9.87623L4.41854 11.2904ZM7.41609 12.8738L10.5828 9.70711L9.16854 8.29289L6.00188 11.4596L7.41609 12.8738ZM6.70898 2.875H8.29232V0.875H6.70898V2.875ZM8.29232 4.04167H6.70898V6.04167H8.29232V4.04167ZM6.70898 4.04167C6.38682 4.04167 6.12565 3.7805 6.12565 3.45833H4.12565C4.12565 4.88507 5.28225 6.04167 6.70898 6.04167V4.04167ZM8.87565 3.45833C8.87565 3.7805 8.61448 4.04167 8.29232 4.04167V6.04167C9.71905 6.04167 10.8757 4.88507 10.8757 3.45833H8.87565ZM8.29232 2.875C8.61448 2.875 8.87565 3.13617 8.87565 3.45833H10.8757C10.8757 2.0316 9.71905 0.875 8.29232 0.875V2.875ZM6.70898 0.875C5.28225 0.875 4.12565 2.0316 4.12565 3.45833H6.12565C6.12565 3.13617 6.38682 2.875 6.70898 2.875V0.875Z"
                fill="#2F3346"
              />
            </svg>
          </div>
          <div
            class="css-j7qwjs"
          >
            <a
              href="/mi-cuenta/mis-compras/123"
            >
              <p
                class="chakra-text css-1oz1cc0"
              >
                Detalle de compra
              </p>
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-vqe3zw"
    >
      <div
        class="css-pk1kxm"
      >
        <p
          class="chakra-text css-j0uba3"
          data-testid="amountTest"
        >
          +
           
          $1.00
        </p>
        <svg
          fill="#04A940"
          height="6"
          transform=""
          viewBox="0 0 10 6"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.24407 0.372871C4.64284 -0.0875905 5.35716 -0.0875904 5.75593 0.372871L8.76318 3.84535C9.32406 4.49299 8.864 5.5 8.00725 5.5L1.99275 5.5C1.136 5.5 0.675944 4.49299 1.23682 3.84535L4.24407 0.372871Z"
          />
        </svg>
      </div>
      <div
        class="css-1xudhsy"
      >
        <p
          class="chakra-text css-1keark"
        >
          Abono
        </p>
      </div>
    </div>
    <div
      class="css-1gjdwbs"
    />
  </div>
</div>
`;
